WordPress/gutenberg
# [Fonts API] Separate BC Layer

Fixes #50073

## What?

The goal is to separate the BC Layer from the Core intended Fonts API code.

With this PR, the BC Layer remains fully functional by invoking the Fonts API's functionality but without making the Fonts API dependent upon the BC Layer.

## Why?

To prepare the API for Core introduction. Why? The BC Layer is Gutenberg specific and will not be merged into Core. Currently this layer is intermingled within the API's code.

## How?
- [X] Adds tests for the BC Layer to ensure functionality continues to work as it did prior to this PR.
- [X] Removes `WP_Webfonts` from the `WP_Fonts` inheritance.
- [X] Changes `wp_webfonts()` back to instantiating `WP_Webfonts()` but injects `wp_fonts()` into it for reuse.
- [X] Moves the deprecated array structure migration code from  `wp_register_fonts()` to `wp_register_webfonts()`.
- [X] Renames the `deprecations` folder to `bc-layer` for readability and better alignment to its purpose.

Originally, `WP_Webfonts` was in the inheritance chain to ensure its public functionality kept working for extenders who may be invoking them. For example, Jetpack's Google Fonts Provider package < v0.5.0 uses `wp_webfonts()` and `WP_Webfonts::get_registered_webfonts()`.

## Testing Instructions

### Automated Tests

### Manually Testing

The key to testing is to compare before and after this PR with emphasis on:
* Does the typography visually looks the same before and after?
* Are the `@font-face` styles in the iframed editor and front-end HTML?

Set up:
1. On your test site, install and activate these 2 plugins:
    * Fonts API Tester using [Jetpack's Google Fonts composer package](https://github.com/ironprogrammer/webfonts-jetpack-test) - installation instructions are found in the [plugins' README.md page](https://github.com/ironprogrammer/webfonts-jetpack-test/blob/master/README.md).
    * [BC Layer Tester plugin](https://github.com/hellofromtonya/fonts-api-bc-layer-tester)
2. A modification is needed in the Fonts API Tester's plugin code. Open its root PHP file in your fav code editor. Comment out the `'Poppins'` font in the `JETPACK_GOOGLE_FONTS_LIST` constant. Why? This is the font used in the BC Layer test plugin.
3. Activate the Gutenberg plugin from this branch.
4. Activate the TT3 theme.

Testing Instructions:
1. Go to the Site Editor > Styles > Typography.
2. For headings, select `Playfair Display` font and set Appearance to `Bold Italic`. Expected behavior: headings typography should change to [look like](https://fonts.google.com/specimen/Playfair+Display?query=playf&preview.text=Mindblown:%20a%20blog%20about%20philosophy.&preview.text_type=custom)
3. For text, select `Poppins` font and set the Appearance to `Regular`. Expected behavior: the paragraph typography should change to [look like](https://fonts.google.com/specimen/Poppins?preview.text=Welcome%20to%20WordPress.%20This%20is%20your%20first%20post.%20Edit%20or%20delete%20it,%20then%20start%20writing!&preview.text_type=custom).
4. Save the styles.
5. Go to the front-end. Do the styles above match?
6. Go to Posts. Open an existing post. Do the styles match?

HTML testing:
The following `style` tags should exist:
```html
<style id="wp-fonts-local">
<style id="wp-fonts-bc-layer">
<style id="wp-fonts-jetpack-google-fonts">
```

The BC Layer style should be like this:
```html
<style id="wp-fonts-bc-layer">
@font-face{font-family:Poppins;font-style:normal;font-weight:400;font-display:fallback;src:local(Poppins), url('/wp-content/plugins/fonts-api-bc-layer-tester/assets/fonts/Poppins/Poppins-Regular.ttf') format('truetype');}@font-face{font-family:Poppins;font-style:italic;font-weight:100 900;font-display:fallback;src:local(Poppins), url('/wp-content/plugins/fonts-api-bc-layer-tester/assets/fonts/Poppins/Poppins-Italic.ttf') format('truetype');}
</style>
````

1. Open your browser's dev tools to inspect the HTML.
2. On the front-end, search for the `style` elements. Are all 3 there?
3. Go to the Site Editor. Again search for the `style` elements. Are all 3 there? (They should be in the `<iframe>`.)
4. Go to Posts. Open an existing post. Again search for the `style` elements. Are all 3 there?

## Screenshots

In the Site Editor
![bclayer-demo](https://user-images.githubusercontent.com/7284611/234966488-bd7ac17c-e1a3-4463-af78-832cdbbe62a8.gif)

On the front-end
![bclayer-fe-demo](https://user-images.githubusercontent.com/7284611/234966433-9327f9e3-bc9e-4baa-a11b-4495a4b07c9f.gif)

In a Post
![bclayer-post-demo](https://user-images.githubusercontent.com/7284611/234966404-24900793-791e-4706-a2cd-20a478e5b519.gif)
